fastcrypto/src/secp256r1/mod.rs

```diff
@@ -1,4 +1,452 @@
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s = <String as serde::Deserialize>::deserialize(deserializer)?;
+        let value = Self::decode_base64(&s).map_err(|e| de::Error::custom(e.to_string()))?;
+        Ok(value)
+    }
+}
+
+impl<'a> From<&'a Secp256r1PrivateKey> for Secp256r1PublicKey {
+    fn from(secret: &'a Secp256r1PrivateKey) -> Self {
+        Secp256r1PublicKey {
+            pubkey: ExternalPublicKey::from(&secret.privkey),
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl SigningKey for Secp256r1PrivateKey {
+    type PubKey = Secp256r1PublicKey;
+    type Sig = Secp256r1Signature;
+    const LENGTH: usize = PRIVATE_KEY_SIZE;
+}
+
+impl ToFromBytes for Secp256r1PrivateKey {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        match ExternalSecretKey::try_from(bytes) {
+            Ok(privkey) => Ok(Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::with_value(<[u8; 32]>::try_from(bytes).unwrap()),
+            }),
+            Err(_) => Err(FastCryptoError::InvalidInput),
+        }
+    }
+}
+
+impl Serialize for Secp256r1PrivateKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_str(&self.encode_base64())
+    }
+}
+
+impl<'de> Deserialize<'de> for Secp256r1PrivateKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+        let value = Self::decode_base64(&s).map_err(|e| de::Error::custom(e.to_string()))?;
+        Ok(value)
+    }
+}
+
+impl AsRef<[u8]> for Secp256r1PrivateKey {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| Ok(self.privkey.to_bytes().into()))
+            .expect("OnceCell invariant violated")
+    }
+}
+
+impl zeroize::Zeroize for Secp256r1PrivateKey {
+    fn zeroize(&mut self) {
+        self.bytes.take().zeroize();
+        // SigningKey from the p256 crate implements zeroize on drop, so we do not need to zeroize it here.
+    }
+}
+
+impl Drop for Secp256r1PrivateKey {
+    fn drop(&mut self) {
+        self.bytes.take().zeroize();
+    }
+}
+
+impl zeroize::ZeroizeOnDrop for Secp256r1PrivateKey {}
+
+impl Serialize for Secp256r1Signature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.as_ref().serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Secp256r1Signature {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let data: Vec<u8> = Vec::deserialize(deserializer)?;
+        <Secp256r1Signature as Signature>::from_bytes(&data)
+            .map_err(|e| de::Error::custom(e.to_string()))
+    }
+}
```

Comment:
the ser/de impl here is OK here - i will clean up with the macro for all crates in a separate pr